servers/jak/shared/common/main/java/com/couchbase/lite/mobiletest/data/KeypathParser.java

```diff
@@ -0,0 +1,171 @@
+                default:
+                    throw new ClientError(errMsg("Unexpected character " + curChar));
+            }
+        }
+    }
+
+    @NonNull
+    private KeyPath.PathElem.Property parseProperty() {
+        // skip and unescaped leading '$'
+        if (curCharClass == CharClass.DOLLAR) { nextChar(); }
+
+        final StringBuilder prop = new StringBuilder();
+        while (true) {
+            switch (curCharClass) {
+                case DOLLAR:
+                case DIGIT:
+                case MISC:
+                    prop.append(curChar);
+                    break;
+                default:
+                    if (prop.length() < 1) { throw new ClientError(errMsg("Empty property")); }
+                    return new KeyPath.PathElem.Property(prop.toString());
+            }
+            nextChar();
+        }
+    }
+
+    @NonNull
+    private KeyPath.PathElem.Index parseIndex() {
+        final StringBuilder idx = new StringBuilder();
+        while (true) {
+            switch (curCharClass) {
+                case DIGIT:
+                    idx.append(curChar);
+                    break;
+                case CBRACKET:
+                    nextChar();
+                    if (idx.length() < 1) { throw new ClientError(errMsg("Empty index")); }
+                    try { return new KeyPath.PathElem.Index(Integer.parseInt(idx.toString())); }
+                    catch (NumberFormatException e) { throw new ClientError(errMsg("Unparsable index"), e); }
+                case EOP:
+                    throw new ClientError(errMsg("Unexpected end of keypath"));
+                default:
+                    throw new ClientError(errMsg("Unexpected character " + curChar + " in index"));
+            }
+            nextChar();
+        }
+    }
+
+    private void nextChar() {
+        nextCh();
+        setChClass();
+    }
+
+    private void pushChar(@NonNull String ch) {
+        curChar = ch;
+        setChClass();
+    }
+
+    private void setChClass() {
+        switch (curChar) {
+            case "":
+                curCharClass = CharClass.EOP;
+                return;
+            case "$":
+                curCharClass = CharClass.DOLLAR;
+                return;
+            case ".":
+                curCharClass = CharClass.DOT;
+                return;
+            case "[":
+                curCharClass = CharClass.OBRACKET;
+                return;
+            case "]":
+                curCharClass = CharClass.CBRACKET;
+                return;
+            case "0":
+            case "1":
+            case "2":
+            case "3":
+            case "4":
+            case "5":
+            case "6":
+            case "7":
+            case "8":
+            case "9":
+                curCharClass = CharClass.DIGIT;
+                return;
+            case "\\":
+                nextCh();
+                if ("".equals(curChar)) { throw new ClientError(errMsg("Dangling escape ('/')")); }
```

Comment:
Your error message slash is the wrong direction.

Reply:
Yep.  Fixed.